websocket/controller.cc

```diff
@@ -35,16 +27,34 @@ future<> controller::start_server()
     }
     _active = true;
     co_await _server.start();
-    co_await _server.invoke_on_all([] (experimental::websocket::server& ws) {
-        ws.register_handler("echo", [] (input_stream<char>& in, output_stream<char>& out) -> future<> {
-            while (true) {
-                temporary_buffer<char> f = co_await in.read();
-                if (f.empty()) {
-                    co_return;
+    co_await _server.invoke_on_all([this] (experimental::websocket::server& ws) {
+        ws.register_handler("cql", [this] (input_stream<char>& in, output_stream<char>& out) -> future<> {
+            auto unix_server_socket = seastar::listen(socket_address{unix_domain_addr{""}});
+            auto unix_addr = unix_server_socket.local_address();
+            auto accept_fut = unix_server_socket.accept();
+            auto sock = co_await connect(unix_addr);
+            auto accept_result = co_await std::move(accept_fut);
+            auto connection = _cql_server.local().inject_connection(unix_addr,
+                    std::move(accept_result.connection), std::move(accept_result.remote_address));
+
+            input_stream<char> cql_in = sock.input();
+            output_stream<char> cql_out = sock.output();
+
+            auto bridge = [] (input_stream<char>& input, output_stream<char>& output) -> future<> {
```

Comment:
The bridge doesn't do any re-coding of the forwarded data. Isn't it then simper/faster/better/hackier to make `generic_server::connection` get constructed with two given `input_stream<char>`s , make such a connection here and inject it directly into the server?

Reply:
Hm, I remember considering and dropping this idea, but I don't remember why and right now it indeed sounds a whole lot simpler than using yet another socket... I'll take another shot at it in v2.

Reply:
Oh I do remember now! `class connection` from generic server creates its own streams from a "connected socket", and the WebSocket handlers accept the streams as references. So in order to be least intrusive with our generic_server, I used a lightweight socket - and then all the code paths for creating streams from the socket remain intact, which reduces the risk of introducing a regression in a quite substantial layer, on top of which all our protocol servers are built.

I'll try investigating if I can "easily" sneak a pair of streams to a connection without building a socket around it first, but I'm afraid it would grow to be a huge, intrusive patch - I remember now that I just dropped after seeing how much of the existing code I need to modify just for an optional web server. Especially that this particular server isn't crucial for performance at all: we would generally assume that clients who use it would be humans sitting in front of a webpage, so a concurrency of 1 and throughput of less than 1 request per second.

Reply:
Maybe patch the `generic_server::connection` to keep two more ~~pointers~~ references on board pointing to the in-line streams by default, but for websocket connection set them to "external" streams?
```patch
diff --git a/generic_server.hh b/generic_server.hh
index 9c3b76ee20..ad160b138b 100644
--- a/generic_server.hh
+++ b/generic_server.hh
@@ -37,8 +37,10 @@ class connection : public boost::intrusive::list_base_hook<> {
 protected:
     server& _server;
     connected_socket _fd;
-    input_stream<char> _read_buf;
-    output_stream<char> _write_buf;
+    input_streap<char> _real_read_buf;
+    input_stream<char>& _read_buf = _real_read_buf;
+    output_stream<char> _real_write_buf;
+    output_stream<char>& _write_buf = _real_write_buf;
     future<> _ready_to_respond = make_ready_future<>();
     seastar::gate _pending_requests_gate;
 
```

Reply:
Yup, that was actually more or less my initial idea, I just felt that I'm going to be bashed in review for introducing a weird layer of indirection just for the sake of an obscure specific use case :) But now that this solution was actually suggested during review, I'm more than happy to give it a try.

Reply:
(although the connected_socket thing will also become optional, so it might be a little more work)